File: src/main/java/AoC2025_04.java
import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2025_04 extends SolutionBase<CharGrid, Long, Long> {

    public static final char ROLL = '@';
    public static final char EMPTY = '.';

    private AoC2025_04(final boolean debug) {
        super(debug);
    }

    public static AoC2025_04 create() {
        return new AoC2025_04(false);
    }

    public static AoC2025_04 createDebug() {
        return new AoC2025_04(true);
    }

    @Override
    protected CharGrid parseInput(final List<String> inputs) {
        return CharGrid.from(inputs);
    }

    private boolean isRemovable(final CharGrid grid, final Cell cell) {
        return grid.getValue(cell) == ROLL
                && grid.getAllNeighbours(cell).filter(n -> grid.getValue(n) == ROLL).count() < 4;
    }

    @Override
    public Long solvePart1(final CharGrid grid) {
        return grid.getCells().filter(cell -> isRemovable(grid, cell)).count();
    }

    @Override
    public Long solvePart2(final CharGrid grid) {
        final Deque<Cell> queue = new ArrayDeque<>();
        grid.getCells().filter(cell -> isRemovable(grid, cell)).forEach(queue::add);
        final Set<Cell> seen = new HashSet<>();
        while (!queue.isEmpty()) {
            final Cell cell = queue.poll();
            if (seen.contains(cell)) {
                continue;
            }
            seen.add(cell);
            grid.setValue(cell, EMPTY);
            grid.getAllNeighbours(cell).filter(n -> isRemovable(grid, n)).forEach(queue::add);
        }
        return (long) seen.size();
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "13"),
        @Sample(method = "part2", input = TEST, expected = "43"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST =
            """
            ..@@.@@@@.
            @@@.@.@.@@
            @@@@@.@.@@
            @.@@@@..@.
            @@.@@@@.@@
            .@@@@@@@.@
            .@.@.@.@@@
            @.@@@.@@@@
            .@@@@@@@@.
            @.@.@@@.@.
            """;
}

File: src/main/python/AoC2025_04.py
#! /usr/bin/env python3
#
# Advent of Code 2025 Day 4
#

import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.geometry import Direction
from aoc.grid import Cell

Input = InputData
Output1 = int
Output2 = int


TEST = """\
..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return input_data

    def part_1(self, inputs: Input) -> Output1:
        grid = list(inputs)
        ans = 0
        for r, row in enumerate(grid):
            for c, ch in enumerate(row):
                if ch == "@":
                    cnt = 0
                    cell = Cell(r, c)
                    for d in Direction.octants():
                        n = cell.at(d)
                        if (
                            0 <= n[0] < len(grid)
                            and 0 <= n[1] < len(row)
                            and grid[n[0]][n[1]] == "@"
                        ):
                            cnt += 1
                    if cnt < 4:
                        ans += 1
        return ans

    def part_2(self, inputs: Input) -> Output2:
        grid = [list(row) for row in inputs]
        ans = 0
        while True:
            new_grid = [list(row) for row in grid]
            for r, row in enumerate(grid):
                for c, ch in enumerate(row):
                    if ch == "@":
                        cnt = 0
                        cell = Cell(r, c)
                        for d in Direction.octants():
                            n = cell.at(d)
                            if (
                                0 <= n[0] < len(grid)
                                and 0 <= n[1] < len(row)
                                and grid[n[0]][n[1]] == "@"
                            ):
                                cnt += 1
                        if cnt < 4:
                            new_grid[r][c] = "."
                            ans += 1
            if new_grid == grid:
                break
            grid = new_grid
        return ans

    @aoc_samples(
        (
            ("part_1", TEST, 13),
            ("part_2", TEST, 43),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2025, 4)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
